06_gpu_and_ml/reinforcement-learning/grpo-verl.py

```diff
@@ -0,0 +1,297 @@
+def prep_dataset() -> None:
+    subprocess.run(
+        [
+            "python",
+            VERL_REPO_PATH / "examples" / "data_preprocess" / "gsm8k.py",
+            "--local_dir",
+            DATA_PATH,
+        ],
+        check=True,
+    )
+
+
+# You can kickoff the dataset download with
+# `modal run <filename.py>::prep_dataset`
+
+# ## Defining a reward function
+
+# In reinforcement learning, we define a reward function for the model.
+# We can define this in a separate file, or in the same file in as in this case that we then pass as an argument to VERL.
+# We use a `default` reward function for GSM8K from the [VERL repo](https://github.com/volcengine/verl/blob/v0.1/verl/utils/reward_score/gsm8k.py), modified to return 1.0 if it's a correct answer and 0 otherwise.
+
+
+def extract_solution(
+    solution_str: str, method: Literal["strict", "flexible"] = "strict"
+) -> Optional[str]:
+    assert method in ["strict", "flexible"]
+
+    if method == "strict":
+        # This also tests the formatting of the model
+        solution = re.search("#### (\\-?[0-9\\.\\,]+)", solution_str)
+        if solution is None:
+            final_answer: Optional[str] = None
+        else:
+            final_answer = solution.group(0)
+            final_answer = (
+                final_answer.split("#### ")[1].replace(",", "").replace("$", "")
+            )
+    elif method == "flexible":
+        answer = re.findall("(\\-?[0-9\\.\\,]+)", solution_str)
+        final_answer: Optional[str] = None
+        if len(answer) == 0:
+            # No reward is there is no answer.
+            pass
+        else:
+            invalid_str: list[str] = ["", "."]
+            # Find the last number that is not '.'
+            for final_answer in reversed(answer):
+                if final_answer not in invalid_str:
+                    break
+    return final_answer
+
+
+# Reward functions need to follow a [predefined signature.](https://verl.readthedocs.io/en/latest/preparation/reward_function.html)
+
+
+def compute_reward(
+    data_source: str, solution_str: str, ground_truth: str, extra_info: dict
+) -> float:
+    answer = extract_solution(solution_str=solution_str, method="strict")
+    if answer is None:
+        return 0.0
+    else:
+        if answer == ground_truth:
+            return 1.0
+        else:
+            return 0.0
+
+
+# We then define constants to pass into VERL during the training run.
+PATH_TO_REWARD_FUNCTION: str = "/root/grpo-verl.py"
+REWARD_FUNCTION_NAME: str = "compute_reward"
+
+# ## Kicking off a training run
+
+# We define some more constants for the training run.
+MODELS_PATH: Path = Path("/models")
+MICROBATCH_SIZE_PER_GPU: int = 16
+MINUTES: int = 60
+HOURS: int = 60 * MINUTES
+DAY: int = 24 * HOURS
+
+# VERL uses Ray under the hood, and creates Ray workers, where each worker can be thought of as a python process, for each of the steps in the RL dataflow pipeline.
```

Comment:
Maybe just `VERL uses Ray under the hood. It creates Ray workers for each step ...`. Why can a worker "be thought of" as a python process? It is or it isn't?